internal/controller/resources/resources.go

```diff
@@ -36,40 +37,25 @@ func IsFailedCleanUp(err error) bool {
 // HandleResourceDeletion ensures the deletion of the nfspvc.
 func HandleResourceDeletion(ctx context.Context, nfspvc danaiov1alpha1.NfsPvc, log logr.Logger, k8sClient client.Client) (error, bool) {
 	if nfspvc.ObjectMeta.DeletionTimestamp != nil {
-		if controllerutil.ContainsFinalizer(&nfspvc, NfsPvcDeletionFinalizer) {
+		if controllerutil.ContainsFinalizer(&nfspvc, utils.NfsPvcDeletionFinalizer) {
 			// check if the pv and the pvc are deleted.
-			pvcDeleted := false
-			pvc := corev1.PersistentVolumeClaim{}
-			if err := k8sClient.Get(ctx, types.NamespacedName{Namespace: nfspvc.Namespace, Name: nfspvc.Name}, &pvc); err != nil {
-				if !errors.IsNotFound(err) {
-					log.Error(err, "unable to get pvc - "+nfspvc.Name)
-					return err, false
-				}
-				pvcDeleted = true
-			}
-			pvDeleted := false
-			pv := corev1.PersistentVolume{}
-			if err := k8sClient.Get(ctx, types.NamespacedName{Name: nfspvc.Name + "-" + nfspvc.Namespace + "-pv"}, &pv); err != nil {
-				if !errors.IsNotFound(err) {
-					log.Error(err, "unable to get pv - "+nfspvc.Name+"-"+nfspvc.Namespace+"-pv")
-					return err, false
-				}
-				pvDeleted = true
+			err, pvcDeleted, pvDeleted := areResourceDeleted(ctx, nfspvc, log, k8sClient)
+			if err != nil {
+				return err, false
 			}
-
 			if pvDeleted && pvcDeleted { // if the pv and the pvc were successfully deleted, remove the finalizer.
-				return removeFinalizer(ctx, nfspvc, log, k8sClient), true
-			} else if !pvDeleted && pvcDeleted { // if only the pvc was deleted successfully, the reconcile function will be triggered again for the pv cleanup.
+				return finalizer.RemoveFinalizer(ctx, nfspvc, log, k8sClient), true // if only the pvc was deleted successfully, the reconcile function will be triggered again for the pv cleanup.
+				// if only the pvc was deleted successfully, the reconcile function will be triggered again for the pv cleanup.
+			} else if !pvDeleted && pvcDeleted {
 				// if the pv still exists, delete it.
-				if err := nfsPvcCleanUp(ctx, nfspvc, log, k8sClient); err != nil {
+				if err := cleanupResources(ctx, nfspvc, log, k8sClient); err != nil {
 					return err, false
 				}
 				pvName := nfspvc.Name + "-" + nfspvc.Namespace + "-pv"
 				return &FailedCleanUpError{Message: "the pv " + pvName + " is not deleted yet"}, false
 			}
-
 			// if the pv and the pvc still exist, delete them.
-			if err := nfsPvcCleanUp(ctx, nfspvc, log, k8sClient); err != nil {
+			if err := cleanupResources(ctx, nfspvc, log, k8sClient); err != nil {
 				return err, false
```

Comment:
This is handled a bit weirdly no? I understand the idea of requeing and not removing the finalizer as long as both the PV and PVC are not deleted but the implementation looks off. We can probably check if they exist but have a deletion timestamp, and if so - requeue and don't attempt to cleanup again.